File: src/main/java/no/nav/fo/veilarbregistrering/db/registrering/BrukerRegistreringRepositoryImpl.java
package no.nav.fo.veilarbregistrering.db.registrering;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import no.nav.fo.veilarbregistrering.besvarelse.Besvarelse;
import no.nav.fo.veilarbregistrering.besvarelse.Stilling;
import no.nav.fo.veilarbregistrering.bruker.AktorId;
import no.nav.fo.veilarbregistrering.bruker.Bruker;
import no.nav.fo.veilarbregistrering.bruker.Foedselsnummer;
import no.nav.fo.veilarbregistrering.registrering.bruker.*;
import no.nav.sbl.sql.DbConstants;
import no.nav.sbl.sql.SqlUtils;
import no.nav.sbl.sql.order.OrderClause;
import no.nav.sbl.sql.where.WhereClause;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.Pageable;
import org.springframework.jdbc.core.JdbcTemplate;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static java.util.Optional.ofNullable;

public class BrukerRegistreringRepositoryImpl implements BrukerRegistreringRepository {

    private final JdbcTemplate db;

    private final static String SYKMELDT_REGISTRERING_SEQ = "SYKMELDT_REGISTRERING_SEQ";
    final static String SYKMELDT_REGISTRERING_ID = "SYKMELDT_REGISTRERING_ID";
    private final static String SYKMELDT_REGISTRERING = "SYKMELDT_REGISTRERING";
    final static String FREMTIDIG_SITUASJON = "FREMTIDIG_SITUASJON";
    final static String TILBAKE_ETTER_52_UKER = "TILBAKE_ETTER_52_UKER";

    private final static String BRUKER_REGISTRERING_SEQ = "BRUKER_REGISTRERING_SEQ";
    private final static String BRUKER_REAKTIVERING_SEQ = "BRUKER_REAKTIVERING_SEQ";
    private final static String BRUKER_REGISTRERING = "BRUKER_REGISTRERING";
    private final static String BRUKER_REAKTIVERING = "BRUKER_REAKTIVERING";
    final static String BRUKER_REGISTRERING_ID = "BRUKER_REGISTRERING_ID";
    private final static String BRUKER_REAKTIVERING_ID = "BRUKER_REAKTIVERING_ID";
    final static String OPPRETTET_DATO = "OPPRETTET_DATO";
    private final static String REAKTIVERING_DATO = "REAKTIVERING_DATO";

    final static String NUS_KODE = "NUS_KODE";
    final static String YRKESPRAKSIS = "YRKESPRAKSIS";
    final static String HAR_HELSEUTFORDRINGER = "HAR_HELSEUTFORDRINGER";
    final static String YRKESBESKRIVELSE = "YRKESBESKRIVELSE";
    final static String KONSEPT_ID = "KONSEPT_ID";
    final static String TEKSTER_FOR_BESVARELSE = "TEKSTER_FOR_BESVARELSE";

    final static String ANDRE_UTFORDRINGER = "ANDRE_UTFORDRINGER";
    final static String BEGRUNNELSE_FOR_REGISTRERING = "BEGRUNNELSE_FOR_REGISTRERING";
    final static String UTDANNING_BESTATT = "UTDANNING_BESTATT";
    final static String UTDANNING_GODKJENT_NORGE = "UTDANNING_GODKJENT_NORGE";
    final static String JOBBHISTORIKK = "JOBBHISTORIKK";

    private final static String AKTOR_ID = "AKTOR_ID";

    public BrukerRegistreringRepositoryImpl(JdbcTemplate db) {
        this.db = db;
    }

    @Override
    public OrdinaerBrukerRegistrering lagre(OrdinaerBrukerRegistrering registrering, Bruker bruker) {
        long id = nesteFraSekvens(BRUKER_REGISTRERING_SEQ);
        Besvarelse besvarelse = registrering.getBesvarelse();
        Stilling stilling = registrering.getSisteStilling();
        String teksterForBesvarelse = tilJson(registrering.getTeksterForBesvarelse());

        SqlUtils.insert(db, BRUKER_REGISTRERING)
                .value(BRUKER_REGISTRERING_ID, id)
                .value(AKTOR_ID, bruker.getAktorId().asString())
                .value("FOEDSELSNUMMER", bruker.getFoedselsnummer().stringValue())
                .value(OPPRETTET_DATO, DbConstants.CURRENT_TIMESTAMP)
                .value(TEKSTER_FOR_BESVARELSE, teksterForBesvarelse)
                // Siste stilling
                .value(YRKESPRAKSIS, stilling.getStyrk08())
                .value(YRKESBESKRIVELSE, stilling.getLabel())
                .value(KONSEPT_ID, stilling.getKonseptId())
                // Besvarelse
                .value(BEGRUNNELSE_FOR_REGISTRERING, besvarelse.getDinSituasjon().toString())
                .value(NUS_KODE, UtdanningUtils.mapTilNuskode(besvarelse.getUtdanning()))
                .value(UTDANNING_GODKJENT_NORGE, besvarelse.getUtdanningGodkjent().toString())
                .value(UTDANNING_BESTATT, besvarelse.getUtdanningBestatt().toString())
                .value(HAR_HELSEUTFORDRINGER, besvarelse.getHelseHinder().toString())
                .value(ANDRE_UTFORDRINGER, besvarelse.getAndreForhold().toString())
                .value(JOBBHISTORIKK, besvarelse.getSisteStilling().toString())
                .execute();

        return hentBrukerregistreringForId(id);
    }

    @Override
    public long lagreSykmeldtBruker(SykmeldtRegistrering bruker, AktorId aktorId) {
        long id = nesteFraSekvens(SYKMELDT_REGISTRERING_SEQ);
        Besvarelse besvarelse = bruker.getBesvarelse();
        String teksterForBesvarelse = tilJson(bruker.getTeksterForBesvarelse());

        SqlUtils.insert(db, SYKMELDT_REGISTRERING)
                .value(SYKMELDT_REGISTRERING_ID, id)
                .value(AKTOR_ID, aktorId.asString())
                .value(OPPRETTET_DATO, DbConstants.CURRENT_TIMESTAMP)
                .value(TEKSTER_FOR_BESVARELSE, teksterForBesvarelse)
                // Besvarelse
                .value(FREMTIDIG_SITUASJON, ofNullable(besvarelse.getFremtidigSituasjon()).map(Enum::toString).orElse(null))
                .value(TILBAKE_ETTER_52_UKER, ofNullable(besvarelse.getTilbakeIArbeid()).map(Enum::toString).orElse(null))
                .value(NUS_KODE, ofNullable(UtdanningUtils.mapTilNuskode(besvarelse.getUtdanning())).orElse(null))
                .value(UTDANNING_BESTATT, ofNullable(besvarelse.getUtdanningBestatt()).map(Enum::toString).orElse(null))
                .value(UTDANNING_GODKJENT_NORGE, ofNullable(besvarelse.getUtdanningGodkjent()).map(Enum::toString).orElse(null))
                .value(ANDRE_UTFORDRINGER, ofNullable(besvarelse.getAndreForhold()).map(Enum::toString).orElse(null))
                .execute();

        return id;
    }

    private static String tilJson(List<TekstForSporsmal> obj) {
        if (obj == null) {
            return "[]";
        }
        try {
            return (new ObjectMapper()).writeValueAsString(obj);
        } catch (JsonProcessingException e) {
            return "[]";
        }
    }

    static List<TekstForSporsmal> tilTeksterForBesvarelse(String json) {
        try {
            TekstForSporsmal[] teksterForBesvarelse = (new ObjectMapper()).readValue(json, TekstForSporsmal[].class);
            return teksterForBesvarelse != null ? Arrays.asList(teksterForBesvarelse) : new ArrayList<>();
        } catch (IOException e) {
            return new ArrayList<>();
        }
    }

    @Override
    public OrdinaerBrukerRegistrering hentBrukerregistreringForId(long brukerregistreringId) {
        return SqlUtils.select(db, BRUKER_REGISTRERING, OrdinaerBrukerRegistreringMapper::map)
                .where(WhereClause.equals(BRUKER_REGISTRERING_ID, brukerregistreringId))
                .column("*")
                .execute();
    }

    @Override
    public OrdinaerBrukerRegistrering hentOrdinaerBrukerregistreringForAktorId(AktorId aktorId) {
        return SqlUtils.select(db, BRUKER_REGISTRERING, OrdinaerBrukerRegistreringMapper::map)
                .where(WhereClause.equals(AKTOR_ID, aktorId.asString()))
                .orderBy(OrderClause.desc(BRUKER_REGISTRERING_ID))
                .limit(1)
                .column("*")
                .execute();
    }

    @Override
    public SykmeldtRegistrering hentSykmeldtregistreringForAktorId(AktorId aktorId) {
        return SqlUtils.select(db, SYKMELDT_REGISTRERING, SykmeldtRegistreringMapper::map)
                .where(WhereClause.equals(AKTOR_ID, aktorId.asString()))
                .orderBy(OrderClause.desc(SYKMELDT_REGISTRERING_ID))
                .limit(1)
                .column("*")
                .execute();
    }

    @Override
    public void lagreReaktiveringForBruker(AktorId aktorId) {
        long id = nesteFraSekvens(BRUKER_REAKTIVERING_SEQ);
        SqlUtils.insert(db, BRUKER_REAKTIVERING)
                .value(BRUKER_REAKTIVERING_ID, id)
                .value(AKTOR_ID, aktorId.asString())
                .value(REAKTIVERING_DATO, DbConstants.CURRENT_TIMESTAMP)
                .execute();
    }

    @Override
    public Bruker hentBrukerTilknyttet(long brukerRegistreringId) {
        String sql = "SELECT FOEDSELSNUMMER, AKTOR_ID FROM BRUKER_REGISTRERING WHERE BRUKER_REGISTRERING_ID = ?";

        return db.queryForObject(sql, new Object[]{brukerRegistreringId}, (rs, i) -> Bruker.of(
                Foedselsnummer.of(rs.getString("FOEDSELSNUMMER")),
                AktorId.of(rs.getString("AKTOR_ID"))
        ));
    }

    private long nesteFraSekvens(String sekvensNavn) {
        return db.queryForObject("select " + sekvensNavn + ".nextval from dual", Long.class);
    }

    @Override
    public Page<ArbeidssokerRegistrertEventDto> findRegistreringByPage(Pageable pageable) {
        String rowCountSql = "SELECT count(1) AS row_count " +
                "FROM BRUKER_REGISTRERING";

        int total = db.queryForObject(rowCountSql, Integer.class);

        String querySql = "SELECT BRUKER_REGISTRERING_ID, AKTOR_ID, BEGRUNNELSE_FOR_REGISTRERING, OPPRETTET_DATO " +
                "FROM BRUKER_REGISTRERING " +
                "ORDER BY BRUKER_REGISTRERING_ID ASC " +
                "OFFSET " + pageable.getOffset() + " ROWS " +
                "FETCH NEXT " + pageable.getPageSize() + " ROWS ONLY";

        List<ArbeidssokerRegistrertEventDto> dto = db.query(
                querySql, (rs, rowNum) -> new ArbeidssokerRegistrertEventDto(
                        rowNum,
                        rs.getLong("BRUKER_REGISTRERING_ID"),
                        AktorId.of(rs.getString("AKTOR_ID")),
                        rs.getString("BEGRUNNELSE_FOR_REGISTRERING"),
                        rs.getTimestamp("OPPRETTET_DATO").toLocalDateTime()
                ));

        return new PageImpl<>(dto, pageable, total);
    }
}

File: src/main/java/no/nav/fo/veilarbregistrering/db/registrering/BrukerRegistreringRepositoryImpl.kt
package no.nav.fo.veilarbregistrering.db.registrering

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import no.nav.fo.veilarbregistrering.besvarelse.*
import no.nav.fo.veilarbregistrering.bruker.AktorId
import no.nav.fo.veilarbregistrering.bruker.Bruker
import no.nav.fo.veilarbregistrering.bruker.Foedselsnummer
import no.nav.fo.veilarbregistrering.db.registrering.RegistreringTilstandRepositoryImpl.Companion.REGISTRERING_TILSTAND
import no.nav.fo.veilarbregistrering.registrering.bruker.BrukerRegistreringRepository
import no.nav.fo.veilarbregistrering.registrering.bruker.OrdinaerBrukerRegistrering
import no.nav.fo.veilarbregistrering.registrering.bruker.TekstForSporsmal
import no.nav.fo.veilarbregistrering.registrering.formidling.Status
import org.springframework.jdbc.core.RowMapper
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate
import java.io.IOException
import java.sql.SQLException
import java.sql.Timestamp
import java.time.LocalDateTime.now

class BrukerRegistreringRepositoryImpl(private val db: NamedParameterJdbcTemplate) : BrukerRegistreringRepository {

    override fun lagre(registrering: OrdinaerBrukerRegistrering, bruker: Bruker): OrdinaerBrukerRegistrering {
        val id = nesteFraSekvens(BRUKER_REGISTRERING_SEQ)
        val besvarelse = registrering.besvarelse
        val stilling = registrering.sisteStilling
        val teksterForBesvarelse = tilJson(registrering.teksterForBesvarelse)

        val params = mapOf(
            "id" to id,
            "aktor_id" to bruker.aktorId.asString(),
            "fnr" to bruker.gjeldendeFoedselsnummer.stringValue(),
            "opprettet" to Timestamp.valueOf(registrering.opprettetDato ?: now()),
            "tekster" to teksterForBesvarelse,
            "yrkespraksis" to stilling.styrk08,
            "yrkesbeskrivelse" to stilling.label,
            "konsept_id" to stilling.konseptId,
            "begrunnelse" to besvarelse.dinSituasjon.toString(),
            "nus_kode" to UtdanningUtils.mapTilNuskode(besvarelse.utdanning),
            "utdanning_godkjent" to besvarelse.utdanningGodkjent.toString(),
            "utdanning_bestatt" to besvarelse.utdanningBestatt.toString(),
            "har_helseutfordringer" to besvarelse.helseHinder.toString(),
            "andre_utfordringer" to besvarelse.andreForhold.toString(),
            "jobbhistorikk" to besvarelse.sisteStilling.toString()
        )
        val sql = "INSERT INTO $BRUKER_REGISTRERING " +
                " (${allColumns.joinToString(", ")})" +
                " VALUES (:id, :aktor_id, :fnr, :opprettet, :tekster, :yrkespraksis, :yrkesbeskrivelse," +
                " :konsept_id, :begrunnelse, :nus_kode, :utdanning_godkjent, :utdanning_bestatt," +
                " :har_helseutfordringer, :andre_utfordringer, :jobbhistorikk)"

        db.update(sql, params)
        return hentBrukerregistreringForId(id)
    }

    override fun hentBrukerregistreringForId(brukerregistreringId: Long): OrdinaerBrukerRegistrering {
        val sql = "SELECT * FROM $BRUKER_REGISTRERING WHERE $BRUKER_REGISTRERING_ID = :id"

        return db.queryForObject(sql, mapOf("id" to brukerregistreringId), registreringMapper)!!
    }

    override fun finnOrdinaerBrukerregistreringForAktorIdOgTilstand(
        aktorId: AktorId,
        tilstander: List<Status>
    ): List<OrdinaerBrukerRegistrering> {
        val params = mapOf(
            "aktor_id" to aktorId.asString(),
            "tilstander" to tilstander.map { it.name }
        )

        val sql = "SELECT * FROM $BRUKER_REGISTRERING" +
                " LEFT JOIN $REGISTRERING_TILSTAND ON " +
                " $REGISTRERING_TILSTAND.${RegistreringTilstandRepositoryImpl.BRUKER_REGISTRERING_ID} = $BRUKER_REGISTRERING.$BRUKER_REGISTRERING_ID" +
                " WHERE $BRUKER_REGISTRERING.$AKTOR_ID = :aktor_id" +
                " AND $REGISTRERING_TILSTAND.${RegistreringTilstandRepositoryImpl.STATUS} in (:tilstander)" +
                " ORDER BY $BRUKER_REGISTRERING.$OPPRETTET_DATO DESC"

        return db.query(sql, params, registreringMapper)
    }

    override fun hentBrukerTilknyttet(brukerRegistreringId: Long): Bruker {
        val sql = "SELECT $FOEDSELSNUMMER, $AKTOR_ID FROM $BRUKER_REGISTRERING WHERE $BRUKER_REGISTRERING_ID = :id"

        return db.queryForObject(sql, mapOf("id" to brukerRegistreringId)) { rs, _ ->
            Bruker.of(
                Foedselsnummer.of(rs.getString("FOEDSELSNUMMER")),
                AktorId.of(rs.getString("AKTOR_ID"))
            )
        }!!
    }

    private fun nesteFraSekvens(sekvensNavn: String): Long {
        return db.queryForObject("SELECT $sekvensNavn.nextval FROM DUAL", noParams, Long::class.java)!!
    }

    companion object {
        private val mapper = jacksonObjectMapper()

        private const val BRUKER_REGISTRERING = "BRUKER_REGISTRERING"
        private const val BRUKER_REGISTRERING_SEQ = "BRUKER_REGISTRERING_SEQ"
        const val FREMTIDIG_SITUASJON = "FREMTIDIG_SITUASJON"
        const val TILBAKE_ETTER_52_UKER = "TILBAKE_ETTER_52_UKER"
        const val BRUKER_REGISTRERING_ID = "BRUKER_REGISTRERING_ID"
        const val OPPRETTET_DATO = "OPPRETTET_DATO"
        const val NUS_KODE = "NUS_KODE"
        const val YRKESPRAKSIS = "YRKESPRAKSIS"
        private const val HAR_HELSEUTFORDRINGER = "HAR_HELSEUTFORDRINGER"
        const val YRKESBESKRIVELSE = "YRKESBESKRIVELSE"
        const val KONSEPT_ID = "KONSEPT_ID"
        const val TEKSTER_FOR_BESVARELSE = "TEKSTER_FOR_BESVARELSE"
        const val ANDRE_UTFORDRINGER = "ANDRE_UTFORDRINGER"
        const val BEGRUNNELSE_FOR_REGISTRERING = "BEGRUNNELSE_FOR_REGISTRERING"
        const val UTDANNING_BESTATT = "UTDANNING_BESTATT"
        const val UTDANNING_GODKJENT_NORGE = "UTDANNING_GODKJENT_NORGE"
        private const val JOBBHISTORIKK = "JOBBHISTORIKK"
        private const val FOEDSELSNUMMER = "FOEDSELSNUMMER"
        const val AKTOR_ID = "AKTOR_ID"

        private val allColumns = listOf(
            BRUKER_REGISTRERING_ID,
            AKTOR_ID,
            FOEDSELSNUMMER,
            OPPRETTET_DATO,
            TEKSTER_FOR_BESVARELSE,
            YRKESPRAKSIS,
            YRKESBESKRIVELSE,
            KONSEPT_ID,
            BEGRUNNELSE_FOR_REGISTRERING,
            NUS_KODE,
            UTDANNING_GODKJENT_NORGE,
            UTDANNING_BESTATT,
            HAR_HELSEUTFORDRINGER,
            ANDRE_UTFORDRINGER,
            JOBBHISTORIKK
        )

        private val registreringMapper = RowMapper<OrdinaerBrukerRegistrering> { rs, _ ->
            try {
                OrdinaerBrukerRegistrering()
                    .setId(rs.getLong(BRUKER_REGISTRERING_ID))
                    .setOpprettetDato(rs.getTimestamp(OPPRETTET_DATO).toLocalDateTime())
                    .setTeksterForBesvarelse(readListOf(rs.getString(TEKSTER_FOR_BESVARELSE)))
                    .setSisteStilling(
                        Stilling()
                            .setStyrk08(rs.getString(YRKESPRAKSIS))
                            .setKonseptId(rs.getLong(KONSEPT_ID))
                            .setLabel(rs.getString(YRKESBESKRIVELSE))
                    )
                    .setBesvarelse(
                        Besvarelse()
                            .setDinSituasjon(DinSituasjonSvar.valueOf(rs.getString(BEGRUNNELSE_FOR_REGISTRERING)))
                            .setUtdanning(UtdanningUtils.mapTilUtdanning(rs.getString(NUS_KODE)))
                            .setUtdanningBestatt(UtdanningBestattSvar.valueOf(rs.getString(UTDANNING_BESTATT)))
                            .setUtdanningGodkjent(UtdanningGodkjentSvar.valueOf(rs.getString(UTDANNING_GODKJENT_NORGE)))
                            .setHelseHinder(HelseHinderSvar.valueOf(rs.getString(HAR_HELSEUTFORDRINGER)))
                            .setAndreForhold(AndreForholdSvar.valueOf(rs.getString(ANDRE_UTFORDRINGER)))
                            .setSisteStilling(SisteStillingSvar.valueOf(rs.getString(JOBBHISTORIKK)))
                    )
            } catch (e: SQLException) {
                throw RuntimeException(e)
            }
        }

        private val noParams = emptyMap<String, Any>()

        private fun tilJson(obj: List<TekstForSporsmal>): String =
            try {
                ObjectMapper().writeValueAsString(obj)
            } catch (e: JsonProcessingException) {
                "[]"
            }

        private inline fun <reified T> readListOf(json: String?) : List<T> =
             try {
                 json?.let {
                     val type = mapper.typeFactory.constructParametricType(List::class.java, T::class.java)
                     jacksonObjectMapper().readValue<List<T>>(it, type)
                 } ?: emptyList()
             } catch (e: IOException) {
                 throw RuntimeException(e)
             }
    }
}